plugins/TwoFactorAuth/TwoFactorAuth.php

```diff
@@ -179,11 +179,15 @@ public function onCreateAppSpecificTokenAuth($returnedValue, $params)
                 // we only return an error when the login/password combo was correct. otherwise you could brute force
                 // auth tokens
                 if (!$authCode) {
-                    http_response_code(401);
+                    if (!headers_sent()) {
+                        http_response_code(401);
+                    }
                     throw new Exception(Piwik::translate('TwoFactorAuth_MissingAuthCodeAPI'));
                 }
                 if (!$twoFa->validateAuthCode($login, $authCode)) {
-                    http_response_code(401);
+                    if (!headers_sent()) {
+                        http_response_code(401);
+                    }
```

Comment:
It seems that we don't do this check for other calls of `http_response_code` in the codebase, is it possible that we need to look into whatever prevents setting this instead?

Reply:
It's actually because our integration tests output text/content before all text is run, which essentially breaks all tests within that might set headers when running the implementation.

( see here https://github.com/matomo-org/matomo/blob/5c0492b68c007474854a417a48f310b2d7816537/tests/PHPUnit/bootstrap.php#L168 )

My solution isn't ideal but at least makes it compatible with the bug that was fixed in 8.3

Reply:
That's correct.
The thing I don't like about my change is that it's in place purely to help with the tests, the problem here is that we have code calling `http_response_code` rather than handling this properly with response objects so output can be handled properly.